pkg/indexer/blockchain/rpcLogStreamer.go

```diff
@@ -0,0 +1,167 @@
+package blockchain
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"go.uber.org/zap"
+)
+
+const (
+	BACKFILL_BLOCKS = 1000
+	// Don't index very new blocks to account for reorgs
+	// Setting to 0 since we are talking about L2s with low reorg risk
+	LAG_FROM_HIGHEST_BLOCK = 0
+	ERROR_SLEEP_TIME       = 100 * time.Millisecond
+)
+
+// The builder that allows you to configure contract events to listen for
+type RpcLogStreamBuilder struct {
+	// All the listeners
+	contractConfigs []contractConfig
+	logger          *zap.Logger
+	rpcUrl          string
+}
+
+func NewRpcLogStreamBuilder(rpcUrl string, logger *zap.Logger) *RpcLogStreamBuilder {
+	return &RpcLogStreamBuilder{rpcUrl: rpcUrl, logger: logger}
+}
+
+func (c *RpcLogStreamBuilder) ListenForContractEvent(fromBlock int, contractAddress common.Address, topics []common.Hash) <-chan types.Log {
+	eventChannel := make(chan types.Log)
+	c.contractConfigs = append(c.contractConfigs, contractConfig{fromBlock, contractAddress, topics, eventChannel})
+	return eventChannel
+}
+
+func (c *RpcLogStreamBuilder) Build() (*RpcLogStreamer, error) {
+	client, err := ethclient.Dial(c.rpcUrl)
+	if err != nil {
+		return nil, err
+	}
+	return NewRpcLogStreamer(client, c.logger, c.contractConfigs), nil
+}
+
+// Struct defining all the information required to filter events from logs
+type contractConfig struct {
+	fromBlock       int
+	contractAddress common.Address
+	topics          []common.Hash
```

Comment:
Are these topics the same as the topics that users would query by? Is this included so that users can query the blockchain directly? For nodes, I imagine they would just firehose everything

Reply:
No, "topics" is an ethereum event term as well. In this case, the topic is actually the hash of the signature of the event itself so that the indexer can pull the right events in contracts that might emit multiple event types.